safehttp/header.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"net/http"
+	"net/textproto"
+)
+
+var disallowedHeaders = map[string]bool{"Set-Cookie": true}
+
+// Header represents the key-value pairs in an HTTP header.
+// The keys will be in canonical form, as returned by
+// textproto.CanonicalMIMEHeaderKey.
+type Header struct {
+	wrappedHeader http.Header
+	immutable     map[string]bool
+}
+
+func newHeader(h http.Header) Header {
+	return Header{wrappedHeader: h, immutable: map[string]bool{}}
+}
+
+// MarkImmutable marks the header with the name `name` as immutable.
+// This header is now read-only. `name` is canonicalized using
+// textproto.CanonicalMIMEHeaderKey first.
+func (h Header) MarkImmutable(name string) {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return
+	}
+	h.immutable[name] = true
+}
+
+// Set sets the header with the name `name` to the value of `value`.
```

Comment:
Style note: Doc comments in Go do not typically use backticks and other formatting things. Instead, try to make the names of the params flow in the text. 

> Set sets the header with the given name to the given value

or something.

You can see how the text is formatted by looking at https://pkg.go.dev/github.com/google/go-safeweb.